File: mobile-app/lib/utils/advanced_cache.dart
/// Cache intelligent multi-niveaux pour mobile
/// L1: Mémoire (rapide), L2: Local storage (persistant)
import 'dart:convert';
import 'package:shared_preferences/shared_preferences.dart';
import 'performance.dart';

class AdvancedCache {
  static final AdvancedCache _instance = AdvancedCache._internal();
  factory AdvancedCache() => _instance;
  AdvancedCache._internal();

  // Cache mémoire (L1)
  final MemoryCache<String, dynamic> _memoryCache = MemoryCache(
    defaultTTL: const Duration(minutes: 5),
  );

  // Statistiques
  int _memoryHits = 0;
  int _storageHits = 0;
  int _misses = 0;
  int _sets = 0;

  /// Obtenir avec cache multi-niveaux
  Future<dynamic> get(String key, {Duration? memoryTTL}) async {
    // Niveau 1: Mémoire
    final memoryValue = _memoryCache.get(key);
    if (memoryValue != null) {
      _memoryHits++;
      return memoryValue;
    }

    // Niveau 2: Local storage
    try {
      final prefs = await SharedPreferences.getInstance();
      final stored = prefs.getString('cache_$key');
      if (stored != null) {
        final data = jsonDecode(stored);
        final timestamp = data['timestamp'] as int;
        final ttl = data['ttl'] as int;
        
        // Vérifier expiration
        if (DateTime.now().millisecondsSinceEpoch - timestamp < ttl) {
          _storageHits++;
          final value = data['value'];
          
          // Mettre en cache mémoire pour prochaine fois
          _memoryCache.set(key, value, ttl: memoryTTL ?? const Duration(minutes: 5));
          
          return value;
        } else {
          // Expiré, supprimer
          await prefs.remove('cache_$key');
        }
      }
    } catch (e) {
      // Erreur storage - continuer sans cache
    }

    _misses++;
    return null;
  }

  /// Mettre en cache avec stratégie intelligente
  Future<void> set(
    String key,
    dynamic value, {
    Duration? ttl,
    Duration? memoryTTL,
    CachePriority priority = CachePriority.normal,
  }) async {
    final cacheTTL = ttl ?? const Duration(hours: 1);
    final memTTL = memoryTTL ?? const Duration(minutes: 5);

    // Toujours mettre en mémoire (accès rapide)
    _memoryCache.set(key, value, ttl: memTTL);

    // Mettre en local storage selon priorité
    if (priority != CachePriority.low) {
      try {
        final prefs = await SharedPreferences.getInstance();
        final data = jsonEncode({
          'value': value,
          'timestamp': DateTime.now().millisecondsSinceEpoch,
          'ttl': cacheTTL.inMilliseconds,
        });
        await prefs.setString('cache_$key', data);
        _sets++;
      } catch (e) {
        // Erreur storage - continuer avec mémoire seulement
      }
    }
  }

  /// Invalider intelligemment
  Future<void> invalidate(String key, {bool pattern = false}) async {
    // Invalider mémoire
    if (pattern) {
      final keysToRemove = _memoryCache.keys.where((k) => k.contains(key)).toList();
      for (final k in keysToRemove) {
        _memoryCache.remove(k);
      }
    } else {
      _memoryCache.remove(key);
    }

    // Invalider storage
    try {
      final prefs = await SharedPreferences.getInstance();
      if (pattern) {
        final allKeys = prefs.getKeys();
        for (final k in allKeys) {
          if (k.startsWith('cache_') && k.contains(key)) {
            await prefs.remove(k);
          }
        }
      } else {
        await prefs.remove('cache_$key');
      }
    } catch (e) {
      // Ignorer erreurs
    }
  }

  /// Précharger des données fréquentes
  Future<int> warmup(List<String> keys, Future<dynamic> Function(String) loaderFn) async {
    int loaded = 0;
    for (final key in keys) {
      try {
        final value = await loaderFn(key);
        if (value != null) {
          await set(key, value, priority: CachePriority.high);
          loaded++;
        }
      } catch (e) {
        // Ignorer erreurs individuelles
      }
    }
    return loaded;
  }

  /// Obtenir les statistiques
  Map<String, dynamic> getStats() {
    final totalHits = _memoryHits + _storageHits;
    final totalRequests = totalHits + _misses;
    final hitRate = totalRequests > 0 ? totalHits / totalRequests : 0.0;

    return {
      'memoryHits': _memoryHits,
      'storageHits': _storageHits,
      'misses': _misses,
      'sets': _sets,
      'hitRate': hitRate,
      'memorySize': _memoryCache.keys.length,
    };
  }

  /// Nettoyer le cache expiré
  Future<void> cleanExpired() async {
    try {
      final prefs = await SharedPreferences.getInstance();
      final allKeys = prefs.getKeys().where((k) => k.startsWith('cache_'));
      final now = DateTime.now().millisecondsSinceEpoch;

      for (final key in allKeys) {
        try {
          final stored = prefs.getString(key);
          if (stored != null) {
            final data = jsonDecode(stored);
            final timestamp = data['timestamp'] as int;
            final ttl = data['ttl'] as int;

            if (now - timestamp > ttl) {
              await prefs.remove(key);
            }
          }
        } catch (e) {
          // Supprimer si erreur de parsing
          await prefs.remove(key);
        }
      }
    } catch (e) {
      // Ignorer erreurs
    }
  }
}

enum CachePriority {
  low,
  normal,
  high,
}

File: backend/utils/advancedCache.js
/**
 * Cache fin et optimisé
 * Cache multi-niveaux avec stratégies intelligentes
 */
const redisCache = require('./redisCache');
const NodeCache = require('node-cache');

class AdvancedCache {
  constructor() {
    // Cache mémoire local (L1)
    this.memoryCache = new NodeCache({
      stdTTL: 300, // 5 minutes par défaut
      checkperiod: 60,
      useClones: false,
    });

    // Statistiques
    this.stats = {
      hits: { memory: 0, redis: 0 },
      misses: 0,
      sets: 0,
    };
  }

  /**
   * Obtenir avec cache multi-niveaux
   */
  async get(key, options = {}) {
    // Niveau 1: Mémoire
    const memoryValue = this.memoryCache.get(key);
    if (memoryValue !== undefined) {
      this.stats.hits.memory++;
      return memoryValue;
    }

    // Niveau 2: Redis
    try {
      const redisValue = await redisCache.get(key);
      if (redisValue !== null) {
        this.stats.hits.redis++;
        // Mettre en cache mémoire pour prochaine fois
        this.memoryCache.set(key, redisValue, options.memoryTTL || 300);
        return redisValue;
      }
    } catch (error) {
      // Redis indisponible - continuer sans cache
      console.warn('Redis cache unavailable, using memory only');
    }

    this.stats.misses++;
    return null;
  }

  /**
   * Mettre en cache avec stratégie intelligente
   */
  async set(key, value, options = {}) {
    const {
      ttl = 3600, // 1 heure par défaut
      memoryTTL = 300, // 5 minutes en mémoire
      priority = 'normal', // normal, high, low
    } = options;

    // Toujours mettre en mémoire (accès rapide)
    this.memoryCache.set(key, value, memoryTTL);

    // Mettre en Redis selon priorité
    if (priority !== 'low') {
      try {
        await redisCache.set(key, value, ttl);
        this.stats.sets++;
      } catch (error) {
        // Redis indisponible - continuer avec mémoire seulement
        console.warn('Redis cache unavailable, using memory only');
      }
    }
  }

  /**
   * Invalider intelligemment
   */
  async invalidate(key, pattern = false) {
    // Invalider mémoire
    if (pattern) {
      const keys = this.memoryCache.keys();
      keys.forEach(k => {
        if (k.includes(key)) {
          this.memoryCache.del(k);
        }
      });
    } else {
      this.memoryCache.del(key);
    }

    // Invalider Redis
    try {
      if (pattern) {
        await redisCache.deletePattern(key);
      } else {
        await redisCache.delete(key);
      }
    } catch (error) {
      // Ignorer erreurs Redis
    }
  }

  /**
   * Précharger des données fréquentes
   */
  async warmup(keys, loaderFn) {
    const results = await Promise.allSettled(
      keys.map(async (key) => {
        const value = await loaderFn(key);
        if (value) {
          await this.set(key, value, { priority: 'high' });
        }
      })
    );

    return results.filter(r => r.status === 'fulfilled').length;
  }

  /**
   * Obtenir les statistiques
   */
  getStats() {
    const totalHits = this.stats.hits.memory + this.stats.hits.redis;
    const totalRequests = totalHits + this.stats.misses;
    const hitRate = totalRequests > 0 ? totalHits / totalRequests : 0;

    return {
      ...this.stats,
      hitRate: parseFloat(hitRate.toFixed(4)),
      memorySize: this.memoryCache.keys().length,
    };
  }
}

module.exports = new AdvancedCache();


